File: legacy/src/llm.mjs
/* LLM 客户端（OpenAI 兼容，如硅基流动 SiliconFlow），零依赖。
   读仓库根 .env（SILICONFLOW_API_KEY / _BASE_URL），或环境变量兜底。 */
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, resolve } from 'node:path';

const here = dirname(fileURLToPath(import.meta.url));
const env = {};
for (const cand of [resolve(here, '..', '..', '.env'), resolve(here, '..', '.env')]) {
  try {
    for (const line of readFileSync(cand, 'utf8').split('\n')) {
      const m = line.match(/^([A-Z_]+)=(.*)$/); if (m && !(m[1] in env)) env[m[1]] = m[2].trim();
    }
  } catch { /* 没有就跳过，靠 process.env */ }
}

export const KEY = env.SILICONFLOW_API_KEY || process.env.SILICONFLOW_API_KEY || process.env.OPENAI_API_KEY;
export const BASE = env.SILICONFLOW_BASE_URL || process.env.SILICONFLOW_BASE_URL || process.env.OPENAI_BASE_URL || 'https://api.siliconflow.cn/v1';
export const MODEL = process.env.LA_MODEL || process.env.SF_MODEL || 'deepseek-ai/DeepSeek-V3';

let nCalls = 0;
export const callCount = () => nCalls;

/* 每次逻辑调用一条记录（含用途/耗时/重试/token）——供 per-run 日志做效果/流程/耗时归因分析。
   nCalls 数的是 HTTP 尝试次数（含重试）；callLog 一条 = 一次 chat()/chatStream() 逻辑调用，attempts 记它内部试了几次。 */
const callLog = [];
export const getCallLog = () => callLog.slice();
export const resetCallLog = () => { callLog.length = 0; };
const promptCharsOf = msgs => msgs.reduce((n, m) => n + ((m && m.content ? String(m.content).length : 0)), 0);

/** OpenAI 兼容 chat。messages=[{role,content}]。jsonMode=true 请求 JSON 对象输出。带重试。
 *  purpose：本次调用的用途标签（plan:skeleton / block:sim / notes …），进 callLog 供归因。 */
export async function chat(messages, { jsonMode = true, temperature = 0.3, model = MODEL, purpose = 'chat' } = {}) {
  if (!KEY) throw new Error('缺 API key：仓库根 .env 里设 SILICONFLOW_API_KEY，或导出 OPENAI_API_KEY');
  const body = { model, messages, temperature };
  if (jsonMode) body.response_format = { type: 'json_object' };
  const promptChars = promptCharsOf(messages);
  const t0 = Date.now();
  let lastErr, attempts = 0;
  const ATTEMPTS = 4;
  for (let attempt = 1; attempt <= ATTEMPTS; attempt++) {
    attempts = attempt;
    try {
      nCalls++;
      const r = await fetch(BASE + '/chat/completions', {
        method: 'POST',
        headers: { Authorization: 'Bearer ' + KEY, 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      });
      if (!r.ok) throw new Error('HTTP ' + r.status + ' ' + (await r.text()).slice(0, 200));
      const d = await r.json();
      const c = d.choices?.[0]?.message?.content;
      if (!c) throw new Error('空响应: ' + JSON.stringify(d).slice(0, 200));
      callLog.push({ seq: callLog.length + 1, purpose, model, stream: false, ms: Date.now() - t0, attempts, ok: true, promptChars, respChars: c.length, promptTokens: d.usage?.prompt_tokens, completionTokens: d.usage?.completion_tokens });
      return c;
    } catch (e) {
      lastErr = e;
      if (attempt < ATTEMPTS) {
        /* 限流/过载/网络 → 更长退避（指数）；其他错误短退避 */
        const rl = /429|rate|too many|529|503|502|overload|timeout|ETIMEDOUT|ECONNRESET|fetch failed/i.test(String(e.message || e));
        await new Promise(res => setTimeout(res, (rl ? 4000 : 1000) * attempt));
      }
    }
  }
  callLog.push({ seq: callLog.length + 1, purpose, model, stream: false, ms: Date.now() - t0, attempts, ok: false, promptChars, err: String(lastErr?.message || lastErr).slice(0, 160) });
  throw lastErr;
}

/** OpenAI 兼容流式 chat。逐 token 调 onDelta(chunk)，返回累积全文。
 *  用于 planner 真流式（token 到达即增量抽取 scene/block 驱动 UI 浮现）。
 *  与 chat() 同源（key/BASE/退避/计数），只多解析 SSE。失败照样重试；不支持流式时调用方回退 chat()。 */
export async function chatStream(messages, { jsonMode = true, temperature = 0.3, model = MODEL, purpose = 'chat' } = {}, onDelta = () => {}) {
  if (process.env.LA_NO_STREAM) throw new Error('LA_NO_STREAM：强制走原子回退（provider 流式不稳时用）');
  if (!KEY) throw new Error('缺 API key：仓库根 .env 里设 SILICONFLOW_API_KEY，或导出 OPENAI_API_KEY');
  const body = { model, messages, temperature, stream: true, stream_options: { include_usage: true } };
  if (jsonMode) body.response_format = { type: 'json_object' };
  const promptChars = promptCharsOf(messages);
  const t0 = Date.now();
  let lastErr, attempts = 0;
  const ATTEMPTS = 4;
  for (let attempt = 1; attempt <= ATTEMPTS; attempt++) {
    attempts = attempt;
    try {
      nCalls++;
      const r = await fetch(BASE + '/chat/completions', {
        method: 'POST',
        headers: { Authorization: 'Bearer ' + KEY, 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      });
      if (!r.ok) throw new Error('HTTP ' + r.status + ' ' + (await r.text()).slice(0, 200));
      if (!r.body || typeof r.body.getReader !== 'function') throw new Error('无流式响应体（环境不支持 ReadableStream）');
      const reader = r.body.getReader();
      const decoder = new TextDecoder();
      let buf = '', full = '', usage = null;
      /* SSE：按空行分事件，每行 data: {json}；末尾 data: [DONE] 结束。跨 chunk 的半行留在 buf。
         include_usage 时，末尾会来一个 choices 为空、带 usage 的块——收下用于 token 计量。 */
      for (;;) {
        const { done, value } = await reader.read();
        if (done) break;
        buf += decoder.decode(value, { stream: true });
        let nl;
        while ((nl = buf.indexOf('\n')) >= 0) {
          const line = buf.slice(0, nl).trim();
          buf = buf.slice(nl + 1);
          if (!line.startsWith('data:')) continue;
          const data = line.slice(5).trim();
          if (data === '[DONE]') continue;
          let piece;
          try { piece = JSON.parse(data); } catch { continue; }   // 忽略半包/心跳行
          if (piece.usage) usage = piece.usage;
          const delta = piece.choices?.[0]?.delta?.content;
          if (delta) { full += delta; try { onDelta(delta, full); } catch { /* UI 回调故障不该中断流 */ } }
        }
      }
      if (!full) throw new Error('流式空响应');
      callLog.push({ seq: callLog.length + 1, purpose, model, stream: true, ms: Date.now() - t0, attempts, ok: true, promptChars, respChars: full.length, promptTokens: usage?.prompt_tokens, completionTokens: usage?.completion_tokens });
      return full;
    } catch (e) {
      lastErr = e;
      if (attempt < ATTEMPTS) {
        const rl = /429|rate|too many|529|503|502|overload|timeout|ETIMEDOUT|ECONNRESET|fetch failed/i.test(String(e.message || e));
        await new Promise(res => setTimeout(res, (rl ? 4000 : 1000) * attempt));
      }
    }
  }
  callLog.push({ seq: callLog.length + 1, purpose, model, stream: true, ms: Date.now() - t0, attempts, ok: false, promptChars, err: String(lastErr?.message || lastErr).slice(0, 160) });
  throw lastErr;
}

/** 从模型输出里抽第一个 JSON（容错去代码围栏）。 */
export function parseJson(txt) {
  let t = String(txt).trim().replace(/^```[a-z]*\n?/i, '').replace(/```$/, '').trim();
  const i = t.indexOf('{'), j = t.lastIndexOf('}');
  if (i < 0 || j < 0) throw new Error('输出里没有 JSON 对象');
  return JSON.parse(t.slice(i, j + 1));
}

/** 简易并发池：对 items 跑 worker，最多 n 个并行。 */
export async function pool(items, n, worker) {
  const out = new Array(items.length);
  let idx = 0;
  async function run() { while (idx < items.length) { const i = idx++; out[i] = await worker(items[i], i); } }
  await Promise.all(Array.from({ length: Math.min(n, items.length) }, run));
  return out;
}

File: notale/utils/llm.py
"""Notale 自包含的 LLM 基础设施。

移植自 lecture-agent（ports/llm.py + adapters/llm/client.py + adapters/llm/fake.py），
已切断对 lecture_agent 包的一切依赖，行为逻辑原样保留。

- HttpxClient —— OpenAI 兼容异步客户端；温度/模型/seed 烘焙在实例上，带指数退避重试。
  确定性优先：默认 temperature=0.0 + 传 seed。api key 从环境变量读，永不入配置。
- FakeClient —— 测试用 in-memory 替身：按顺序/purpose 返回罐装响应，不联网。
  证明调用方只依赖接口：单测传它即可端到端跑，无需真 key。
"""

from __future__ import annotations

import asyncio
import os
import re
from collections import deque
from typing import Any, Protocol, TypedDict, runtime_checkable

import httpx

# ---------------------------------------------------------------------------
# 类型与接口
# ---------------------------------------------------------------------------


class Message(TypedDict):
    role: str
    content: str


@runtime_checkable
class LLMClient(Protocol):
    async def complete(
        self,
        messages: list[Message],
        *,
        json_mode: bool = True,
        purpose: str = "chat",
    ) -> str:
        """一次 OpenAI 兼容 chat，返回文本。json_mode 请求 JSON 对象输出。

        purpose 是用途标签（plan:skeleton / block:sim …），供归因日志，不影响语义。
        """
        ...


# ---------------------------------------------------------------------------
# 真调客户端
# ---------------------------------------------------------------------------

_RATE_LIMIT = re.compile(
    r"429|rate|too many|529|503|502|overload|timeout|ETIMEDOUT|ECONNRESET", re.I
)


class HttpxClient:
    def __init__(
        self,
        *,
        base_url: str = "https://api.siliconflow.cn/v1",
        model: str = "deepseek-ai/DeepSeek-V4-Flash",
        temperature: float = 0.0,
        seed: int | None = 0,
        api_key_env: str = "SILICONFLOW_API_KEY",
        attempts: int = 4,
        timeout: float = 600.0,  # reasoning 模型单调用可能 >120s（沿用 lecture-agent 对该模型的 pin）
        proxy: str | None = None,
        extra_body: dict[str, Any] | None = None,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.model = model
        self.temperature = temperature
        self.seed = seed
        self.api_key_env = api_key_env
        self.attempts = attempts
        self.timeout = timeout
        # 境外端点（如 Gemini 评委）需显式走翻墙代理；国内端点留 None=直连。
        self.proxy = proxy
        # 任意 OpenAI 兼容额外参数透传（enable_thinking / thinking / reasoning_effort / max_tokens …）。
        # 普适："关思考"等 per-model 差异全交配置，不进代码。
        self.extra_body = dict(extra_body or {})
        # 累计 token（供实验 harness 归因；仅真调计数，命中缓存的调用不计）
        self.usage: dict[str, int] = {
            "prompt_tokens": 0,
            "completion_tokens": 0,
            "reasoning_tokens": 0,
            "total_tokens": 0,
            "calls": 0,
        }

    @property
    def signature(self) -> str:
        """标识"用什么参数产出"，供录制盒命名空间隔离（换模型/温度不撞 fixture）。"""
        return f"{self.model}@t{self.temperature}@s{self.seed}"

    def _api_key(self) -> str:
        key = os.environ.get(self.api_key_env) or os.environ.get("OPENAI_API_KEY")
        if not key:
            raise RuntimeError(f"缺 API key：设环境变量 {self.api_key_env} 或 OPENAI_API_KEY")
        return key

    async def _post_message(self, body: dict[str, Any]) -> dict[str, Any]:
        """POST /chat/completions，带退避重试，返回 choices[0].message 对象。"""
        headers = {"Authorization": f"Bearer {self._api_key()}", "Content-Type": "application/json"}
        last_err: Exception | None = None
        # trust_env=False：忽略环境 *_PROXY（本地翻墙代理不路由国内端点，且 SOCKS 需额外依赖）。
        # self.proxy 显式给定时（境外评委端点）走该 http 代理；否则直连。
        async with httpx.AsyncClient(
            timeout=self.timeout, trust_env=False, proxy=self.proxy
        ) as client:
            for attempt in range(1, self.attempts + 1):
                try:
                    resp = await client.post(
                        f"{self.base_url}/chat/completions", headers=headers, json=body
                    )
                    if resp.status_code != 200:
                        raise RuntimeError(f"HTTP {resp.status_code} {resp.text[:200]}")
                    data = resp.json()
                    self._record_usage(data.get("usage"))
                    message: dict[str, Any] = data.get("choices", [{}])[0].get("message", {})
                    return message
                except Exception as e:  # noqa: BLE001 - 统一重试
                    last_err = e
                    if attempt < self.attempts:
                        backoff = (4.0 if _RATE_LIMIT.search(str(e)) else 1.0) * attempt
                        await asyncio.sleep(backoff)
        assert last_err is not None
        raise last_err

    def _record_usage(self, usage: dict[str, Any] | None) -> None:
        """累加一次真调的 token 用量（OpenAI 兼容 usage 字段；缺字段兜 0）。"""
        if not usage:
            return
        self.usage["prompt_tokens"] += int(usage.get("prompt_tokens", 0) or 0)
        self.usage["completion_tokens"] += int(usage.get("completion_tokens", 0) or 0)
        self.usage["total_tokens"] += int(usage.get("total_tokens", 0) or 0)
        details = usage.get("completion_tokens_details") or {}
        self.usage["reasoning_tokens"] += int(details.get("reasoning_tokens", 0) or 0)
        self.usage["calls"] += 1

    def _base_body(self) -> dict[str, Any]:
        body: dict[str, Any] = {
            "model": self.model,
            "messages": [],
            "temperature": self.temperature,
        }
        if self.seed is not None:
            body["seed"] = self.seed
        body.update(self.extra_body)  # 透传关思考/reasoning_effort/max_tokens 等
        return body

    async def complete(
        self, messages: list[Message], *, json_mode: bool = True, purpose: str = "chat"
    ) -> str:
        body = self._base_body()
        body["messages"] = messages
        if json_mode:
            body["response_format"] = {"type": "json_object"}
        message = await self._post_message(body)
        content = message.get("content")
        if not content:
            raise RuntimeError("空响应")
        return str(content)

# ---------------------------------------------------------------------------
# 测试替身
# ---------------------------------------------------------------------------


class FakeClient:
    """测试用 in-memory LLMClient。按顺序/purpose 返回罐装响应，不联网。"""

    def __init__(
        self,
        queue: list[str] | None = None,
        by_purpose: dict[str, str] | None = None,
        default: str = "{}",
    ) -> None:
        self._queue: deque[str] = deque(queue or [])
        self._by_purpose = by_purpose or {}
        self._default = default
        self.calls: list[tuple[str, list[Message]]] = []

    async def complete(
        self, messages: list[Message], *, json_mode: bool = True, purpose: str = "chat"
    ) -> str:
        self.calls.append((purpose, messages))
        if purpose in self._by_purpose:
            return self._by_purpose[purpose]
        if self._queue:
            return self._queue.popleft()
        return self._default
